### PR TITLE
Implements EPC for nim-mode in nimsuggest

### PR DESCRIPTION
**Pull request by reactormonk**

### CODE REVIEW

**Comment by Araq** [comment]
You need to rebase this stuff.

**Comment by reactormonk** [comment]
Made a proper merge commit instead.

**Comment by Varriount** [comment]
Uh, are you sure? There's still an awful lot of changes...

**Comment by reactormonk** [comment]
@Varriount indeed. I implemented a sexp parser/generator and rewrote most of nimsuggest.

File: compiler/suggest.nim [hunk 1]
```diff
@@ -11,61 +11,88 @@
 
 # included from sigmatch.nim
 
-import algorithm, sequtils
+import algorithm, sequtils, strutils, ast, msgs, options, renderer,
+  types, docgen, lexer, semdata, astalgo, idents, sigmatch, lookups
```
Comment (Araq) [inline compiler/suggest.nim]: If suggest.nim is included from sigmatch, why do you import sigmatch here? And in fact, I'm sure these deps shouldn't be here.

File: compiler/suggest.nim [hunk 2]
```diff
@@ -15,57 +15,83 @@ import algorithm, sequtils
...
-proc symToStr(s: PSym, isLocal: bool, section: string, li: TLineInfo): string = 
-  result = section
-  result.add(sep)
+proc symToSuggest(s: PSym, isLocal: bool, section: string, li: TLineInfo): Suggest = 
+  result.section = parseIdeCmd(section)
   if optIdeTerse in gGlobalOptions:
     if s.kind in routineKinds:
-      result.add renderTree(s.ast, {renderNoBody, renderNoComments,
-                                    renderDocComments, renderNoPragmas})
+      result.symkind = parseEnum[TSymKind](renderTree(s.ast, {renderNoBody, renderNoComments,
```
Comment (Araq) [inline compiler/suggest.nim]: WTF?

File: compiler/suggest.nim [hunk 3]
```diff
@@ -15,57 +15,83 @@ import algorithm, sequtils
...
+                                    renderDocComments, renderNoPragmas}))
     else:
-      result.add s.name.s
-    result.add(sep)
-    result.add(toFullPath(li))
-    result.add(sep)
-    result.add($toLinenumber(li))
-    result.add(sep)
-    result.add($toColumn(li))
+      result.symkind = parseEnum[TSymKind](s.name.s)
```
Comment (Araq) [inline compiler/suggest.nim]: WTF?

File: compiler/suggest.nim (hunk 2 above)
Comment (reactormonk) [inline compiler/suggest.nim]: Copied over from https://github.com/Araq/Nim/blob/devel/compiler/suggest.nim#L32-L33

File: compiler/suggest.nim (hunk 3 above)
Comment (reactormonk) [inline compiler/suggest.nim]: Pulled over from https://github.com/Araq/Nim/blob/devel/compiler/suggest.nim#L53

**Comment by dom96** [comment]
What's EPC?

**Comment by reactormonk** [comment]
https://github.com/kiwanami/emacs-epc

**Comment by dom96** [comment]
Shouldn't this be implemented on top of nimsuggest?

**Comment by reactormonk** [comment]
I prefer nimsuggest <-> emacs over nimsuggest <-> epc server <-> emacs

**Comment by dom96** [comment]
This also adds an extra module to the stdlib which we have to maintain. I think that makes it an even more likely candidate for a Nimble package.

But it's merged now so meh. At the very least though, add the ``sexp`` module to the docs and change logs.

**Comment by reactormonk** [comment]
@dom96 How do I add it to the docs? And I'm not sure it warrants a mention in the change logs, I hope we can export it as nimble package sooner rather than later.

**Comment by dom96** [comment]
We have a "Library additions" section in the changelogs, so it is warranted.

Add it to https://github.com/Araq/Nim/blob/devel/doc/lib.txt and overview.txt (I think). It seems the process changed recently, this used to have the list of docs to be generated: https://github.com/Araq/Nim/blob/devel/compiler/installer.ini#L45, now I guess it goes through overview.txt.

**Pull request merged**